File: xide-standard-plugins/xide-logs-plugin/src/main/kotlin/org/xast/xide/logs_plugin/ConsoleRedirectStream.kt
package org.xast.xide.logs_plugin

import java.io.OutputStream
import java.nio.charset.StandardCharsets

class ConsoleRedirectStream(val view: LogsView, val error: Boolean) : OutputStream() {
    private val buffer = StringBuilder()

    override fun write(b: Int) {
        write(byteArrayOf(b.toByte()), 0, 1)
    }

    override fun write(b: ByteArray, off: Int, len: Int) {
        var text = String(b, off, len, StandardCharsets.UTF_8)

        for (c in text) {
            when (c) {
                '\n' -> {
                    view.addLine(buffer.toString(), error)
                    buffer.setLength(0)
                }
                '\r' -> { 
                    // ignore
                }
                else -> buffer.append(c)
            }
        }
    }
}

File: xide-standard-plugins/xide-logs-plugin/src/main/scala/org/xast/xide/logs_plugin/ConsoleRedirectStream.scala
package org.xast.xide.logs_plugin

import java.io.OutputStream
import java.nio.charset.StandardCharsets

class ConsoleRedirectStream(
    val view: LogsView, 
    val error: Boolean
) extends OutputStream():
    
    private val buffer = StringBuilder()

    override def write(b: Int) = write(Array(b.toByte), 0, 1)

    override def write(b: Array[Byte], off: Int, len: Int) =
        val text = new String(b, off, len, StandardCharsets.UTF_8)

        text.foreach:
            case '\n' =>
                view.addLine(buffer.toString, error)
                buffer.setLength(0)
                
            case '\r' => // ignore

            case c => buffer.append(c)